File: backend/src/models/TelemetryReading.model.js
// src/models/TelemetryReading.model.js
// Time-series telemetry readings ingested from smart plugs.
// Includes a 90-day TTL index for automatic data retention management.

const mongoose = require('mongoose');

const TelemetryReadingSchema = new mongoose.Schema(
  {
    // References
    plugId: {
      type: String,
      required: true,
      index: true,
    },
    applianceId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Appliance',
      default: null,
      index: true,
    },
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true,
      index: true,
    },

    // ── Electrical measurements ──────────────────────────────────────────────
    wattage: {
      type: Number,
      required: true,
      min: 0,
    },
    voltage: {
      type: Number,
      default: null, // ~230V AC for India
    },
    current: {
      type: Number,
      default: null, // Amperes
    },
    powerFactor: {
      type: Number,
      default: null, // 0.0 – 1.0
    },

    // ── Anomaly metadata ─────────────────────────────────────────────────────
    isAnomaly: {
      type: Boolean,
      default: false,
      index: true,
    },
    anomalyScore: {
      type: Number,
      default: null, // Z-score value
    },
    anomalyReason: {
      type: String,
      default: null, // Human-readable explanation
    },

    // ── Reading timestamp ────────────────────────────────────────────────────
    // We keep a dedicated `timestamp` field (vs relying on createdAt) so that
    // external systems or backfills can supply the correct measurement time.
    timestamp: {
      type: Date,
      required: true,
      default: Date.now,
      index: true,
    },
  },
  {
    // Disable auto `createdAt`/`updatedAt` to avoid redundancy with `timestamp`
    timestamps: false,
  },
);

// ── Compound indexes for efficient dashboard queries ─────────────────────────
TelemetryReadingSchema.index({ plugId: 1, timestamp: -1 });
TelemetryReadingSchema.index({ userId: 1, isAnomaly: 1, timestamp: -1 });
TelemetryReadingSchema.index({ applianceId: 1, timestamp: -1 });

// ── TTL index: auto-delete readings older than 90 days ───────────────────────
TelemetryReadingSchema.index(
  { timestamp: 1 },
  { expireAfterSeconds: 60 * 60 * 24 * 90 }, // 90 days
);

module.exports = mongoose.model('TelemetryReading', TelemetryReadingSchema);

File: frontend/lib/feature/smart_plug/models/telemetry_reading_model.dart
// lib/feature/smart_plug/models/telemetry_reading_model.dart

class TelemetryReading {
  final String id;
  final String plugId;
  final double wattage;
  final double? voltage;
  final double? current;
  final double? powerFactor;
  final bool isAnomaly;
  final double? anomalyScore;
  final String? anomalyReason;
  final DateTime timestamp;

  const TelemetryReading({
    required this.id,
    required this.plugId,
    required this.wattage,
    this.voltage,
    this.current,
    this.powerFactor,
    required this.isAnomaly,
    this.anomalyScore,
    this.anomalyReason,
    required this.timestamp,
  });

  factory TelemetryReading.fromMap(Map<String, dynamic> map) {
    return TelemetryReading(
      id:           map['id'] as String? ?? map['_id'] as String? ?? '',
      plugId:       map['plugId'] as String? ?? '',
      wattage:      (map['wattage'] as num?)?.toDouble() ?? 0.0,
      voltage:      (map['voltage'] as num?)?.toDouble(),
      current:      (map['current'] as num?)?.toDouble(),
      powerFactor:  (map['powerFactor'] as num?)?.toDouble(),
      isAnomaly:    map['isAnomaly'] as bool? ?? false,
      anomalyScore: (map['anomalyScore'] as num?)?.toDouble(),
      anomalyReason: map['anomalyReason'] as String?,
      timestamp: DateTime.tryParse(map['timestamp'] as String? ?? '') ??
          DateTime.now(),
    );
  }
}
